### ISSUE TITLE
Integrals threading error messages

### ISSUE THREAD

**Issue by susilehtola** [psi4/psi4#596]
Running larger calculations with scf_type pk gives me lots and lots of error messages
```
$ psi4
thread already exists.
thread already exists.
thread already exists.
```
and so on.

These seem to arise from src/psi4/libpsio/aio_handler.cc. I'm wondering if these really are errors, if they should be printed, and whether stdout instead of the log file is the right place to do it.

**Comment by dgasmith** [comment]
Can you supply an input. @jgonthier any comments on this?

**Comment by susilehtola** [comment]
```
molecule {
0 1
symmetry c1
Mn -0.0128264866 -0.0002779912 -0.1091894544
O  -1.1037647823  1.0246639195 -1.2821459116
O  -0.4319309790 -1.5910632792 -1.0423314985
N   0.7243918073  1.6205280457  0.7709992473
N   1.3008767173 -0.9287768925  1.0635839133
C  -1.2886186814  2.2834208702 -1.2181338272
C  -0.7050865946  3.1821062059 -0.3604271375
C   0.3534920334  2.8195308082  0.5207059889
C   0.2177419816 -2.6945497951 -0.9662726987
C   1.2181378357 -3.0250490115 -0.0961995911
C   1.7180430400 -2.1229929449  0.8899899444
C   1.9393113446  1.3525171542  1.5367699106
C   1.8333722479 -0.0604962603  2.1055467704
H   0.8908035853  3.6241374839  1.0039698993
H   2.5024254412 -2.4854165868  1.5399007635
H   2.7858100041  1.4098930461  0.8661311665
H   1.1490336627 -0.0652673733  2.9502674019
H   2.0638511115  2.0769335299  2.3353274192
H   2.7993768711 -0.4085469833  2.4534782102
H  -1.9892347032  2.6686747336 -1.9454961675
H  -0.0986239324 -3.4441353556 -1.6764363233
Cl  1.8821942342  0.3098852687 -1.6599082362
O  -1.1395010178 -0.1703680057  0.9263551082
H  -0.9312248148  4.2251864392 -0.4696082573
H   1.6695378989 -3.9953914470 -0.1633815245
}

# Run SCF
set basis 6-31G*
set scf_type pk
set soscf true
scf_e, scf_wfn = energy('scf', return_wfn=True)
```

**Comment by jgonthier** [comment]
Mmh, how many parallel processes do you have ? aio_handler.cc spawns a new thread that is specifically for writing the integrals to disk. When there is no writing left to do, the thread should die, and a new one is generated when further writing is needed.

Here it looks like it's trying to generate the thread when it already exists... Or maybe the threads are not properly dying, and it gets to the maximum number of threads.

Does your computation seem to proceed normally in spite of the error message ?

**Comment by susilehtola** [comment]
The computation appears to proceed normally. I believe Psi4 was using only one thread (I did not supply a -n argument), although I have set $OMP_NUM_THREADS=4...

**Comment by jgonthier** [comment]
Ok, thank you Susi.

@loriab I see that you have added a synchronization step in AIOHandler::write, see former commit id c205621. Do you remember the reason for that ? As far as I can remember, the way AIOHandler is designed we should never need a synchronization here.

@jturney From the git history, the error message Susi is seeing was added by you during KTB. Do you remember why you put it there ?

**Comment by loriab** [comment]
I remember the episode but not the exact error (though it was in sapt and cubeprop tests). Something about pybind's behavior being less forgiving than boost's. Also psi4 modules behaved differently wrt how much control they took upon themselves vs giving to the AIOHandler lib. I think I standardized that aspect a bit. Good news is that I think you can adjust the synchronization again, so long as the tests pass (since it was broken test cases before that prompted the change). Or the inversion may have spontaneously healed the problem.

**Comment by jgonthier** [comment]
Just a little update: unless jturney has an opposition, I think we can safely delete the error message. I'm now trying to clean up the synchronization issues. I found the problem with cubeprop, and I'm working on the errors in the sapt tests, which seem to be even more subtle.

**Comment by jgonthier** [comment]
@loriab, @jturney : Problem fixed. Deleting synchronization in AIOHandler::write, I saw the errors you mentioned, Lori. The error in cubeprop was caused by the threads not being joined before they were destroyed (Boost automatically called detach() in the thread destructors). After adding synchronization everywhere, the cubeprop error disappears.
Then, sapt2, sapt3 and sapt6 had errors. These ones were caused by threads being joined twice when the user called synchronize on AIOHandler. That is now fixed by having synchronize() check wether the threads are joinable or not.

I believe threads are now properly managed by AIOHandler.
@susilehtola : the error message you saw will not appear after PR #609 is accepted. Thanks for reporting it!
